packages/expo-device/android/src/main/java/expo/modules/device/DeviceModule.java

```diff
@@ -0,0 +1,385 @@
+package expo.modules.device;
+
+import android.app.Activity;
+import android.content.Context;
+import android.content.pm.PackageManager;
+import android.net.wifi.WifiInfo;
+import android.net.wifi.WifiManager;
+import android.os.Bundle;
+import android.util.Log;
+import android.os.Build;
+import android.os.Environment;
+import android.provider.Settings;
+import android.support.v4.content.ContextCompat;
+import android.support.v4.app.ActivityCompat;
+import android.app.KeyguardManager;
+import android.os.StatFs;
+import android.telephony.TelephonyManager;
+import android.app.ActivityManager;
+import android.app.UiModeManager;
+import android.view.WindowManager;
+import android.util.DisplayMetrics;
+import android.content.res.Configuration;
+import android.Manifest;
+
+import org.unimodules.core.ExportedModule;
+import org.unimodules.core.ModuleRegistry;
+import org.unimodules.core.Promise;
+import org.unimodules.core.interfaces.ActivityProvider;
+import org.unimodules.core.interfaces.ExpoMethod;
+import org.unimodules.core.interfaces.RegistryLifecycleListener;
+
+import java.math.BigInteger;
+import java.net.InetAddress;
+import java.net.NetworkInterface;
+import java.nio.ByteOrder;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+
+public class DeviceModule extends ExportedModule implements RegistryLifecycleListener {
+  private static final String NAME = "ExpoDevice";
+  private static final String TAG = DeviceModule.class.getSimpleName();
+  private final int REQUEST_READ_PHONE_STATE = 1;
+
+  private ModuleRegistry mModuleRegistry;
+  private Context mContext;
+  private WifiInfo wifiInfo;
+  private ActivityProvider mActivityProvider;
+  private Activity mActivity;
+  private String mPhoneNumber;
+  DeviceType deviceType;
+
+  public DeviceModule(Context context) {
+    super(context);
+    mContext = context;
+    deviceType = this.getDeviceType(context);
+  }
+
+  public enum DeviceType {
+    HANDSET("Handset"),
+    TABLET("Tablet"),
+    TV("Tv"),
+    UNKNOWN("Unknown");
+
+    private final String value;
+
+    DeviceType(String value) {
+      this.value = value;
+    }
+
+    public String getValue() {
+      return value;
+    }
+  }
+
+
+  private WifiInfo getWifiInfo() {
+    if (this.wifiInfo == null) {
+      WifiManager manager = (WifiManager) mContext.getApplicationContext().getSystemService(Context.WIFI_SERVICE);
+      this.wifiInfo = manager.getConnectionInfo();
+    }
+    return this.wifiInfo;
+  }
+
+  @Override
+  public String getName() {
+    return NAME;
+  }
+
+  @Override
+  public void onCreate(ModuleRegistry moduleRegistry) {
+    mModuleRegistry = moduleRegistry;
+    mActivityProvider = moduleRegistry.getModule(ActivityProvider.class);
+    mActivity = mActivityProvider.getCurrentActivity();
+    mPhoneNumber = getPhoneNumber();
```

Comment:
This is a "phoneNumber" constant in the module and it requires a read_phone_state permission. For such dangerous permission, it would ask for such permission at runtime. So in order to get the permission before the user tries to get the phone number, I need to ask for the permission when initialized.

Reply:
If I understand you correctly, calling `getPhoneNumber()` triggers permissions dialog, right? I have doubts whether asking for the permission once the application starts is the best way of doing this—often apps want to first explain to users why they need certain personal data and only then ask/not ask for it.

So because
- it requires a runtime permission dialog and we want to show it only when the user asks
- theoretically phone number could change during runtime,

1. `phoneNumber` can't be in constants,
2. `phoneNumber` can't be a cached variable,
3. we may want to expose `getPhoneNumberAsync` and explicitly state in the documentation it will trigger permissions dialog

Reply:
Doing a getPhoneNumberAsync() is indeed more logical.

Reply:
I took a closer look at the logic, but I think it would be like this (reference from https://developer.android.com/training/permissions/requesting): 

1. The user calls `getPhoneNumberAsync()`
2. Check for permission of `READ_PHONE_STATE`
3. If not granted, request the permission
1) granted after request, get `mPhoneNumber` in current state.
2) not granted, `mPhoneNumber` = informative message(permission is not granted)
4.If already granted, `mPhoneNumber` would fetch the phoneNumber value again.

So in this way, keeping a private variable `mPhoneNumber` seems reasonable since we need to store the phoneNumber value after permission request grant the permission and need to access it through `getPhoneNumberAsync()`. And the `mPhoneNumber` value would update each time since we fetch the value when the user calls the `getPhoneNumberAsync()`.

I'm not sure if this makes any sense or not.